File: rewrite-core/src/main/java/com/netflix/rewrite/tree/Type.java
/*
 * Copyright 2020 the original authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.netflix.rewrite.tree;

import com.fasterxml.jackson.annotation.*;
import com.koloboke.collect.map.hash.HashObjObjMaps;
import com.koloboke.collect.set.hash.HashObjSet;
import com.koloboke.collect.set.hash.HashObjSets;
import com.netflix.rewrite.internal.lang.Nullable;
import lombok.Data;
import lombok.Getter;
import lombok.Setter;
import lombok.With;

import java.io.Serializable;
import java.util.*;

import static java.util.Collections.emptyList;
import static java.util.Comparator.comparing;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;

@JsonIdentityInfo(generator = ObjectIdGenerators.IntSequenceGenerator.class, property = "@ref")
@JsonTypeInfo(use = JsonTypeInfo.Id.MINIMAL_CLASS, property = "@c")
public interface Type extends Serializable {
    boolean deepEquals(@Nullable Type type);

    @Data
    class MultiCatch implements Type {
        private final List<Type> throwableTypes;

        @Override
        public boolean deepEquals(@Nullable Type type) {
            return type instanceof MultiCatch &&
                    TypeUtils.deepEquals(throwableTypes, ((MultiCatch) type).throwableTypes);
        }
    }

    /**
     * Reduces memory and CPU footprint when deep class insight isn't necessary, such as
     * for the type parameters of a Type.Class
     */
    @Data
    class ShallowClass implements Type {
        private final String fullyQualifiedName;

        @Override
        public boolean deepEquals(Type type) {
            return type instanceof ShallowClass &&
                    fullyQualifiedName.equals(((ShallowClass) type).fullyQualifiedName);
        }
    }

    @Getter
    class Class implements Type {
        // there shouldn't be too many distinct types represented by the same fully qualified name
        private static final Map<String, HashObjSet<Class>> flyweights = HashObjObjMaps.newMutableMap();

        public static final Class OBJECT = build("java.lang.Object");

        private final String fullyQualifiedName;
        private final List<Var> members;
        private final List<Type> typeParameters;
        private final List<Type> interfaces;

        @Nullable
        private final Class supertype;

        private Class(String fullyQualifiedName, List<Var> members, List<Type> typeParameters, List<Type> interfaces, @Nullable Class supertype) {
            this.fullyQualifiedName = fullyQualifiedName;
            this.members = members;
            this.typeParameters = typeParameters;
            this.interfaces = interfaces;
            this.supertype = supertype;
        }

        /**
         * Build a class type only from the class' fully qualified name. Since we are not providing any member, type parameter,
         * interface, or supertype information, this fully qualified name could potentially match on more than one version of
         * the class found in the type cache. This method will simply pick one of them, because there is no way of selecting
         * between the versions of the class based solely on the fully qualified class name.
         *
         * @param fullyQualifiedName The fully qualified name of the class to build
         * @return Any class found in the type cache
         */
        public static Class build(String fullyQualifiedName) {
            return build(fullyQualifiedName, emptyList(), emptyList(), emptyList(), null, true);
        }

        @JsonCreator
        public static Class build(@JsonProperty("fullyQualifiedName") String fullyQualifiedName,
                                  @JsonProperty("members") List<Var> members,
                                  @JsonProperty("typeParameters") List<Type> typeParameters,
                                  @JsonProperty("interfaces") List<Type> interfaces,
                                  @JsonProperty("supertype") @Nullable Class supertype,
                                  boolean relaxedClassTypeMatching) {

            // when class type matching is NOT relaxed, the variants are the various versions of this fully qualified
            // name, where equality is determined by whether the supertype hierarchy and members through the entire
            // supertype hierarchy are equal
            var test = new Class(fullyQualifiedName,
                    members.stream().sorted(comparing(Var::getName)).collect(toList()),
                    typeParameters, interfaces, supertype);

            synchronized (flyweights) {
                var variants = flyweights.computeIfAbsent(fullyQualifiedName, fqn -> HashObjSets.newMutableSet());
                return (!relaxedClassTypeMatching ? variants.stream().filter(v -> v.deepEquals(test)) : variants.stream())
                        .findAny()
                        .orElseGet(() -> {
                            variants.add(test);
                            return test;
                        });
            }
        }

        @JsonIgnore
        public String getClassName() {
            return Arrays.stream(fullyQualifiedName.split("\\."))
                    .dropWhile(part -> Character.isLowerCase(part.charAt(0)))
                    .collect(joining("."));
        }

        @JsonIgnore
        public String getPackageName() {
            return Arrays.stream(fullyQualifiedName.split("\\."))
                    .takeWhile(part -> part.length() > 0 && !Character.isUpperCase(part.charAt(0)))
                    .collect(joining("."));
        }

        @JsonIgnore
        public boolean isAssignableFrom(@Nullable Type.Class clazz) {
            return clazz != null && (this == OBJECT ||
                    this.fullyQualifiedName.equals(clazz.fullyQualifiedName) ||
                    isAssignableFrom(clazz.getSupertype()) ||
                    clazz.getInterfaces().stream().anyMatch(i -> i instanceof Class && isAssignableFrom((Class) i))
            );
        }

        @JsonIgnore
        public List<Type.Var> getVisibleSupertypeMembers() {
            List<Type.Var> members = new ArrayList<>();
            if(supertype != null) {
                supertype.getMembers().stream()
                        .filter(member -> !member.hasFlags(Flag.Private))
                        .forEach(members::add);
                members.addAll(supertype.getVisibleSupertypeMembers());
            }
            return members;
        }

        @Override
        public boolean deepEquals(@Nullable Type type) {
            if (!(type instanceof Class)) {
                return false;
            }

            Class c = (Class) type;
            return fullyQualifiedName.equals(c.fullyQualifiedName) &&
                    TypeUtils.deepEquals(members, c.members) &&
                    TypeUtils.deepEquals(supertype, c.supertype) &&
                    TypeUtils.deepEquals(typeParameters, c.typeParameters);
        }
    }

    @Data
    class Cyclic implements Type {
        private final String fullyQualifiedName;

        @Override
        public boolean deepEquals(Type type) {
            return this.equals(type);
        }
    }

    @Data
    class Var implements Type {
        private final String name;

        @Nullable
        private final Type type;

        private final Set<Flag> flags;

        public boolean hasFlags(Flag... test) {
            return Arrays.stream(test).allMatch(flags::contains);
        }

        @Override
        public boolean deepEquals(@Nullable Type type) {
            if (!(type instanceof Var)) {
                return false;
            }

            Var v = (Var) type;
            return name.equals(v.name) && TypeUtils.deepEquals(this.type, v.type) &&
                    flags.equals(v.flags);
        }
    }

    @Getter
    class Method implements Type {
        private static final Map<Class, Map<String, Set<Method>>> flyweights = HashObjObjMaps.newMutableMap();

        @With
        private final Class declaringType;

        private final String name;
        private final Signature genericSignature;
        private final Signature resolvedSignature;
        private final List<String> paramNames;

        @With
        private final Set<Flag> flags;

        private Method(Class declaringType, String name, Signature genericSignature, Signature resolvedSignature, List<String> paramNames, Set<Flag> flags) {
            this.declaringType = declaringType;
            this.name = name;
            this.genericSignature = genericSignature;
            this.resolvedSignature = resolvedSignature;
            this.paramNames = paramNames;
            this.flags = flags;
        }

        @JsonCreator
        public static Method build(@JsonProperty("declaringType") Class declaringType,
                                   @JsonProperty("name") String name,
                                   @JsonProperty("genericSignature") Signature genericSignature,
                                   @JsonProperty("resolvedSignature") Signature resolvedSignature,
                                   @JsonProperty("paramNames") List<String> paramNames,
                                   @JsonProperty("flags") Set<Flag> flags) {
            var test = new Method(declaringType, name, genericSignature, resolvedSignature, paramNames, flags);

            synchronized (flyweights) {
                Set<Method> methods = flyweights
                        .computeIfAbsent(declaringType, dt -> HashObjObjMaps.newMutableMap())
                        .computeIfAbsent(name, n -> HashObjSets.newMutableSet());

                return methods
                        .stream()
                        .filter(m -> m.deepEquals(test))
                        .findAny()
                        .orElseGet(() -> {
                            methods.add(test);
                            return test;
                        });
            }
        }

        @Data
        public static class Signature implements Serializable {
            @Nullable
            private final Type returnType;

            private final List<Type> paramTypes;
        }

        private static boolean signatureDeepEquals(@Nullable Signature s1, @Nullable Signature s2) {
            return s1 == null ? s2 == null : s2 != null &&
                    TypeUtils.deepEquals(s1.returnType, s2.returnType) &&
                    TypeUtils.deepEquals(s1.paramTypes, s2.paramTypes);
        }

        public boolean hasFlags(Flag... test) {
            return Arrays.stream(test).allMatch(flags::contains);
        }

        @Override
        public boolean deepEquals(Type type) {
            if (!(type instanceof Method)) {
                return false;
            }

            Method m = (Method) type;
            return paramNames.equals(m.paramNames) &&
                    flags.equals(m.flags) &&
                    declaringType.deepEquals(m.declaringType) &&
                    signatureDeepEquals(genericSignature, m.genericSignature) &&
                    signatureDeepEquals(resolvedSignature, m.resolvedSignature);
        }
    }

    @Data
    class GenericTypeVariable implements Type {
        private final String fullyQualifiedName;

        @Nullable
        private final Class bound;

        @Override
        public boolean deepEquals(Type type) {
            if (!(type instanceof GenericTypeVariable)) {
                return false;
            }

            GenericTypeVariable generic = (GenericTypeVariable) type;
            return fullyQualifiedName.equals(generic.fullyQualifiedName) &&
                    TypeUtils.deepEquals(bound, generic.bound);
        }
    }

    @Data
    class Array implements Type {
        private final Type elemType;

        @Override
        public boolean deepEquals(Type type) {
            return type instanceof Array && elemType.deepEquals(((Array) type).elemType);
        }
    }

    enum Primitive implements Type {
        Boolean, Byte, Char, Double, Float, Int, Long, Short, Void, String, None, Wildcard, Null;

        public static Primitive fromKeyword(String keyword) {
            switch (keyword) {
                case "boolean":
                    return Boolean;
                case "byte":
                    return Byte;
                case "char":
                    return Char;
                case "double":
                    return Double;
                case "float":
                    return Float;
                case "int":
                    return Int;
                case "long":
                    return Long;
                case "short":
                    return Short;
                case "void":
                    return Void;
                case "String":
                    return String;
                case "":
                    return None;
                case "*":
                    return Wildcard;
                case "null":
                    return Null;
                default:
                    throw new IllegalArgumentException("Invalid primitive type " + keyword);
            }
        }

        @JsonIgnore
        public String getKeyword() {
            switch (this) {
                case Boolean:
                    return "boolean";
                case Byte:
                    return "byte";
                case Char:
                    return "char";
                case Double:
                    return "double";
                case Float:
                    return "float";
                case Int:
                    return "int";
                case Long:
                    return "long";
                case Short:
                    return "short";
                case Void:
                    return "void";
                case String:
                    return "String";
                case Wildcard:
                    return "*";
                case Null:
                    return "null";
                case None:
                default:
                    return "";
            }
        }

        @Override
        public boolean deepEquals(Type type) {
            return this == type;
        }
    }
}

File: rewrite-core/src/main/kotlin/com/netflix/rewrite/ast/Type.kt
/**
 * Copyright 2016 Netflix, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.netflix.rewrite.ast

import com.fasterxml.jackson.annotation.*
import com.koloboke.collect.map.hash.HashObjObjMap
import com.koloboke.collect.map.hash.HashObjObjMaps
import com.koloboke.collect.set.hash.HashObjSet
import com.koloboke.collect.set.hash.HashObjSets
import com.netflix.rewrite.ast.Type.Companion.deepEquals
import java.io.Serializable

@JsonIdentityInfo(generator = ObjectIdGenerators.IntSequenceGenerator::class, property = "@ref")
@JsonTypeInfo(use = JsonTypeInfo.Id.MINIMAL_CLASS, include = JsonTypeInfo.As.PROPERTY, property = "@c")
sealed class Type: Serializable {

    companion object {
        internal fun Type?.deepEquals(t: Type?): Boolean = when (this) {
            null -> t == null
            is Array -> t is Array && this.deepEquals(t)
            is Class -> t is Class && this.deepEquals(t)
            is Cyclic -> this == t
            is GenericTypeVariable -> t is GenericTypeVariable && this.deepEquals(t)
            is Method -> t is Method && this.deepEquals(t)
            is Primitive -> t is Primitive && this == t
            is Var -> t is Var && this.deepEquals(t)
            is ShallowClass -> t is ShallowClass && fullyQualifiedName == t.fullyQualifiedName
            is MultiCatchType -> t is MultiCatchType && throwableTypes.size == t.throwableTypes.size &&
                    throwableTypes.indices.all { i -> throwableTypes[i].deepEquals(t.throwableTypes[i]) }
        }
    }

    data class MultiCatchType(val throwableTypes: List<Type>): Type()

    /**
     * Reduces memory and CPU footprint when deep class insight isn't necessary, such as
     * for the type parameters of a Type.Class
     */
    data class ShallowClass(val fullyQualifiedName: String): Type()

    data class Class private constructor(val fullyQualifiedName: String,
                                         val members: List<Var>, // will always be sorted by name by build(..)
                                         val supertype: Class?,
                                         val typeParameters: List<Type>,
                                         val interfaces: List<Type>): Type() {

        override fun toString(): String = fullyQualifiedName

        fun className() = fullyQualifiedName.split('.').dropWhile { it[0].isLowerCase() }.joinToString(".")

        override fun hashCode() = fullyQualifiedName.hashCode()

        fun packageName(): String {
            fun packageNameInternal(fqn: String): String {
                if(!fqn.contains("."))
                    return ""

                val subName = fqn.substringBeforeLast(".")
                return if (subName.substringAfterLast('.').first().isUpperCase()) {
                    packageNameInternal(subName)
                } else {
                    subName
                }
            }

            return packageNameInternal(fullyQualifiedName)
        }

        fun deepEquals(c: Class?): Boolean {
            if(c == null || fullyQualifiedName != c.fullyQualifiedName)
                return false

            if(members.size != c.members.size)
                return false
            members.indices.forEach {
                if(!members[it].deepEquals(c.members[it])) return@deepEquals false
            }

            if(!supertype.deepEquals(c.supertype))
                return false

            if(typeParameters.size != c.typeParameters.size)
                return false
            typeParameters.indices.forEach {
                if(!typeParameters[it].deepEquals(c.typeParameters[it])) return@deepEquals false
            }

            return true
        }

        companion object {
            var classVersionDependentComparison = true

            // there shouldn't be too many distinct types represented by the same fully qualified name
            val flyweights = HashObjObjMaps.newMutableMap<String, HashObjSet<Class>>()

            @JvmStatic @JsonCreator
            fun build(fullyQualifiedName: String,
                      members: List<Var> = emptyList(),
                      supertype: Class? = null,
                      typeParameters: List<Type> = emptyList(),
                      interfaces: List<Type> = emptyList()): Type.Class {
                // the variants are the various versions of this fully qualified name, where equality is determined by
                // whether the supertype hierarchy and members through the entire supertype hierarchy are equal
                val test = Class(fullyQualifiedName, members.sortedBy { it.name }, supertype, typeParameters, interfaces)

                return synchronized(flyweights) {
                    val variants = flyweights.getOrPut(fullyQualifiedName) {
                        HashObjSets.newMutableSet<Class>(arrayOf(Class(fullyQualifiedName, members, supertype, typeParameters, interfaces)))
                    }

                    (if(classVersionDependentComparison) variants.find { it.deepEquals(test) } else variants.firstOrNull()) ?: {
                        variants.add(test)
                        test
                    }()
                }
            }
        }
    }

    data class Cyclic(val fullyQualifiedName: String) : Type()

    data class Method private constructor(val declaringType: Class,
                                          val name: String,
                                          val genericSignature: Signature?,
                                          val resolvedSignature: Signature?,
                                          val paramNames: List<String>?,
                                          val flags: List<Flag>): Type() {

        fun hasFlags(vararg test: Flag) = test.all { flags.contains(it) }

        data class Signature(val returnType: Type?, val paramTypes: List<Type>): Serializable

        internal fun deepEquals(method: Method?): Boolean {
            if(method == null)
                return false

            return declaringType.deepEquals(method.declaringType) &&
                    genericSignature.deepEquals(method.genericSignature) &&
                    resolvedSignature.deepEquals(method.resolvedSignature) &&
                    flags.all { method.flags.contains(it) } &&
                    paramNames?.equals(method.paramNames) ?: method.paramNames == null
        }

        companion object {
            val flyweights = HashObjObjMaps.newMutableMap<Class, HashObjObjMap<String, HashObjSet<Method>>>()

            private fun Signature?.deepEquals(signature: Signature?): Boolean {
                if(this == null)
                    return signature == null
                if(signature == null)
                    return false

                return returnType.deepEquals(signature.returnType) &&
                        paramTypes.size == signature.paramTypes.size &&
                        (paramTypes.isEmpty() || paramTypes.zip(signature.paramTypes).map { (p1, p2) -> p1.deepEquals(p2) }.reduce(Boolean::and))
            }

            @JvmStatic @JsonCreator
            fun build(declaringType: Class, name: String, genericSignature: Signature?, resolvedSignature: Signature?,
                      paramNames: List<String>, flags: List<Flag>): Type.Method {
                val test = Method(declaringType, name, genericSignature, resolvedSignature, paramNames, flags)

                return synchronized(flyweights) {
                    val methods = flyweights
                            .getOrPut(declaringType, { HashObjObjMaps.newMutableMap(mapOf(name to HashObjSets.newMutableSet(listOf(test)))) })
                            .getOrPut(name, { HashObjSets.newMutableSet(listOf(test)) })

                    methods.find { m ->
                        paramNames == m.paramNames &&
                                genericSignature.deepEquals(m.genericSignature) &&
                                resolvedSignature.deepEquals(m.resolvedSignature) &&
                                flags.all { m.flags.contains(it) }
                    } ?: {
                        methods.add(test)
                        test
                    }()
                }
            }
        }
    }

    data class GenericTypeVariable(val fullyQualifiedName: String, val bound: Class?): Type() {
        internal fun deepEquals(generic: GenericTypeVariable?): Boolean =
                generic != null && fullyQualifiedName == generic.fullyQualifiedName && bound.deepEquals(generic.bound)
    }

    data class Array(val elemType: Type): Type() {
        internal fun deepEquals(array: Array?): Boolean =
                array != null && elemType.deepEquals(array.elemType)
    }

    data class Primitive(val keyword: String): Type() {
        companion object {
            val Boolean = Primitive("boolean")
            val Byte = Primitive("byte")
            val Char = Primitive("char")
            val Double = Primitive("double")
            val Float = Primitive("float")
            val Int = Primitive("int")
            val Long = Primitive("long")
            val Short = Primitive("short")
            val Void = Primitive("void")
            val String = Primitive("String")
            val None = Primitive("")
            val Wildcard = Primitive("*")
            val Null = Primitive("null")

            fun build(keyword: String): Primitive = when(keyword) {
                "boolean" -> Boolean
                "byte" -> Byte
                "char" -> Char
                "double" -> Double
                "float" -> Float
                "int" -> Int
                "long" -> Long
                "short" -> Short
                "void" -> Void
                "String" -> String
                "" -> None
                "*" -> Wildcard
                "null" -> Null
                else -> throw IllegalArgumentException("Invalid primitive ordinal")
            }
        }
    }

    data class Var(val name: String, val type: Type?, val flags: List<Flag>): Type() {
        fun hasFlags(vararg test: Flag) = test.all { flags.contains(it) }

        override fun hashCode() = name.hashCode()

        internal fun deepEquals(v: Var): Boolean =
                name == v.name && type.deepEquals(v.type) && flags.all { v.flags.contains(it) }
    }
}

fun Type?.asClass(): Type.Class? = when(this) {
    is Type.Class -> this
    else -> null
}

fun Type?.asArray(): Type.Array? = when(this) {
    is Type.Array -> this
    else -> null
}

fun Type?.asGeneric(): Type.GenericTypeVariable? = when(this) {
    is Type.GenericTypeVariable -> this
    else -> null
}

fun Type?.asMethod(): Type.Method? = when(this) {
    is Type.Method -> this
    else -> null
}

fun Type?.hasElementType(fullyQualifiedName: String): Boolean = when(this) {
    is Type.Array -> this.elemType.hasElementType(fullyQualifiedName)
    is Type.Class -> this.fullyQualifiedName == fullyQualifiedName
    is Type.GenericTypeVariable -> this.fullyQualifiedName == fullyQualifiedName
    else -> false
}

fun Type?.asPrimitive(): Type.Primitive? = when(this) {
    is Type.Primitive -> this
    else -> null
}